node/core/chain-selection/src/lib.rs

```diff
@@ -0,0 +1,757 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Implements the Chain Selection Subsystem.
+
+use polkadot_primitives::v1::{BlockNumber, Hash, Header, ConsensusLog};
+use polkadot_subsystem::{
+	Subsystem, SubsystemContext, SubsystemResult, SubsystemError, SpawnedSubsystem,
+	OverseerSignal, FromOverseer,
+	messages::{ChainSelectionMessage, ChainApiMessage},
+	errors::ChainApiError,
+};
+
+use parity_scale_codec::Error as CodecError;
+use futures::channel::oneshot;
+
+use std::collections::HashMap;
+use std::time::{UNIX_EPOCH, SystemTime};
+
+const LOG_TARGET: &str = "parachain::chain-selection";
+
+type Weight = u64;
+type Timestamp = u64;
+
+#[derive(Debug, Clone)]
+enum Approval {
+	// Approved
+	Approved,
+	// Unapproved but not stagnant
+	Unapproved,
+	// Unapproved and stagnant.
+	Stagnant,
+}
+
+impl Approval {
+	fn is_stagnant(&self) -> bool {
+		match *self {
+			Approval::Stagnant => true,
+			_ => false,
+		}
+	}
+}
+
+#[derive(Debug, Clone)]
+struct ViabilityCriteria {
+	// Whether this block has been explicitly reverted by one of its descendants.
+	explicitly_reverted: bool,
+	// `None` means approved. `Some` means unapproved.
+	approval: Approval,
+	earliest_unviable_ancestor: Option<Hash>,
```

Comment:
Comment is not entirely clear, what would `approval` be set to in that case?

Reply:
outdated comment when `Approval` was an `Option<bool>`